Add {In,Out}_channel.isatty

nojb opened the pull request:
Simply expose the existing primitive in `Sys` (the same code used to implement `Unix.isatty`). Motivated by discussion in #11127. cc @dbuenzli

nojb committed:
Add Sys.isatty

nojb committed:
Changes

dbuenzli commented:
Wonderful! 

But now that I see the signature, my idea to put in in `Sys` was misleading. `isatty` is can be applied to both input and output fds. 

That looks an excellent opportunity to use the new `{In,Out}_channel` modules.

nojb committed:
Move to In_channel, Out_channel

nojb commented:
> Wonderful!
> 
> But now that I see the signature, my idea to put in in `Sys` was misleading. `isatty` is can be applied to both input and output fds.
> 
> That looks an excellent opportunity to use the new `{In,Out}_channel` modules.

Good point, amended.

dbuenzli commented on stdlib/out_channel.mli:
```diff
@@ -158,3 +158,9 @@ val set_buffered : t -> bool -> unit
     {!set_buffered}). *)
+
+val isatty : t -> bool
+(** Returns [true] if the given channel refers to a terminal or console
```
You could have retained the documentation style of `is_buffered` (i.e. `isatty oc` is `true` if the channel `oc` refers…). But I'm nitpicking.

nojb replied:
Wish granted, thanks.

nojb committed:
Wording

avsm commented on stdlib/in_channel.mli:
```diff
@@ -150,3 +150,9 @@ val set_binary_mode : t -> bool -> unit
+
+val isatty : t -> bool
+(** [isatty ic] is [true] if [ic] refers to a terminal or console window,
+    [false] otherwise.
```
A nitpick is that the [Windows documentation for isatty](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/isatty?view=msvc-170) refers to:
>  a character device (a terminal, console, printer, or serial port).

Perhaps it's more accurate to refer to a character device instead, but I haven't checked to see what isatty on Linux does does if you give it a serial port device.

nojb replied:
> I haven't checked to see what isatty on Linux does does if you give it a serial port device.

It returns `false`. Personally I would keep the existing somewhat vague docstring.

nojb commented:
Thanks @gasche for the approval. Per the rule, this PR needs a second approval from a core dev before it can be merged, any takers?

gasche commented:
Note: I wondered how this new feature of channels would be adapted for the proposal of making the channel type externally-extensible ([RFC: modular IO](https://github.com/ocaml/RFCs/pull/19)). My conclusion was that it adds yet another small wrinkle to the proposal, but nothing worse than what we already have with, say, `fseek`, so it's probably fine.

nojb commented:
Friendly ping (still in need of a second approval for this easy PR). Maybe @Octachron?

Octachron commented:
I might have some time once the stdlib audit and concurrency-safety documentation for 5.0 is done.

dbuenzli commented:
I'm sure @Octachron has enough things on his plate. But could we maybe get approval from a less busy core dev team member to get this in ?

gasche commented:
I opened the thread to join to help, and then I realized that I am the first approver...

Octachron reviewed (approved):
The `isatty` function seems like a fine addition to me:
- the function already exists and is merely exposed
- it is an important property of the underlying channel
- it should relatively straightforward for custom channels to implement this function

nojb committed:
Update Changes, docstrings

nojb commented:
> The `isatty` function seems like a fine addition to me:

Thanks for the approval. I updated the "since" docstring to `5.1.0` and moved the Changes entry to the corresponding section. Planning to merge once CI passes.
